// @cedros/pay-react-native/src/ecommerce/types.ts
export type ProductImage = {
  url: string;
  alt?: string;
};

/**
 * A value within a variation type (e.g., "Medium" for Size, "Blue" for Color)
 */
export type VariationValue = {
  id: string;
  label: string;
  /** Optional: parent value ID for hierarchical dependencies */
  parentValueId?: string;
};

/**
 * A variation type definition (e.g., "Size", "Color")
 */
export type VariationType = {
  id: string;
  name: string;
  /** Display order (0 = first selector) */
  displayOrder: number;
  values: VariationValue[];
};

/**
 * Product-level variation configuration
 */
export type ProductVariationConfig = {
  variationTypes: VariationType[];
};

export type ProductVariant = {
  id: string;
  title: string;
  options: Record<string, string>; // e.g. { Size: 'M', Color: 'Black' }
  /** References to VariationValue IDs that make up this variant */
  optionValueIds?: string[];
  price?: number; // major units
  compareAtPrice?: number; // major units
  inventoryStatus?: 'in_stock' | 'low' | 'out_of_stock' | 'backorder';
  /** Actual inventory count (when tracked) */
  inventoryQuantity?: number;
  /** Stock keeping unit */
  sku?: string;
  /** Variant-specific images */
  images?: ProductImage[];
  /** Whether this variant was auto-generated */
  autoGenerated?: boolean;
};

export type Product = {
  id: string;
  slug: string;
  title: string;
  description: string;
  images: ProductImage[];
  price: number; // major units
  currency: string;
  tags: string[];
  categoryIds: string[];
  /** Variation type definitions for this product */
  variationConfig?: ProductVariationConfig;
  variants?: ProductVariant[];
  inventoryStatus?: 'in_stock' | 'low' | 'out_of_stock' | 'backorder';
  /** When present, inventory is tracked and finite */
  inventoryQuantity?: number;
  compareAtPrice?: number; // major units
  attributes?: Record<string, string | number | boolean>;
  shippingProfile?: 'physical' | 'digital';
  checkoutRequirements?: {
    email?: 'none' | 'optional' | 'required';
    name?: 'none' | 'optional' | 'required';
    phone?: 'none' | 'optional' | 'required';
    shippingAddress?: boolean;
    billingAddress?: boolean;
  };
  fulfillment?: {
    type: 'digital_download' | 'shipping' | 'service';
    notes?: string;
  };
};

export type Category = {
  id: string;
  slug: string;
  name: string;
  description?: string;
  parentId?: string;
  bannerImageUrl?: string;
};

export type CartItem = {
  productId: string;
  variantId?: string;
  qty: number;
  unitPrice: number; // major units, snapshot
  currency: string;
  titleSnapshot: string;
  imageSnapshot?: string;
  // Optional provider mapping for payment boundary
  paymentResource?: string;
  metadata?: Record<string, string>;
  /** Inventory hold ID if holds are enabled */
  holdId?: string;
  /** When the inventory hold expires (ISO datetime) */
  holdExpiresAt?: string;
};

export type CartSnapshot = {
  items: CartItem[];
  promoCode?: string;
};

export type OrderLineItem = {
  title: string;
  qty: number;
  unitPrice: number;
  currency: string;
  imageUrl?: string;
};

/** Order status values matching backend */
export type OrderStatus =
  | 'created'
  | 'paid'
  | 'processing'
  | 'fulfilled'
  | 'shipped'
  | 'delivered'
  | 'cancelled'
  | 'refunded';

export type Order = {
  id: string;
  createdAt: string;
  status: OrderStatus;
  total: number;
  currency: string;
  items: OrderLineItem[];
  receiptUrl?: string;
  invoiceUrl?: string;
  /** Payment source/method: 'stripe', 'x402', or 'credits' */
  source?: 'stripe' | 'x402' | 'credits';
  /** Purchase/transaction ID (Stripe payment intent, blockchain tx hash, or credits tx ID) */
  purchaseId?: string;
  /** Customer email (if collected) */
  customerEmail?: string;
  /** Customer name (if collected) */
  customerName?: string;
};

export type CheckoutMode = 'none' | 'minimal' | 'shipping' | 'full';

export type CustomerInfo = {
  email?: string;
  name?: string;
  phone?: string;
  notes?: string;
  shippingAddress?: Address;
  billingAddress?: Address;
};

export type Address = {
  line1: string;
  line2?: string;
  city: string;
  state?: string;
  postalCode: string;
  country: string;
};

export type ShippingMethod = {
  id: string;
  label: string;
  price: number; // major units
  currency: string;
  detail?: string;
};

export type ListResult<T> = {
  items: T[];
  page: number;
  pageSize: number;
  total?: number;
  hasNextPage?: boolean;
};
